File: rq1/scripts/table_17.py
"""
reproduce_table17.py — UK-residency threshold sensitivity (k=1..5).

60 models: 2 groups × 2 definitions × 5 thresholds × 3 milestones.

Input:  rq1/data/table17_weekly_series.csv
Output: rq1/data/table17_reproduced.csv

Usage:
  python rq1/reproduce_table17.py
"""
import os, warnings
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"
warnings.filterwarnings("ignore")
import pandas as pd

if not hasattr(pd.DataFrame, "applymap"):
    pd.DataFrame.applymap = pd.DataFrame.map
from causalimpact import CausalImpact

DATA_DIR = os.path.join("rq1", "data")

MILESTONES = {
    "Oct2023": pd.Timestamp("2023-10-23"),
    "Mar2025": pd.Timestamp("2025-03-17"),
    "Jul2025": pd.Timestamp("2025-07-21"),
}

GROUPS = ["VPN", "Politics"]
DEFINITIONS = ["Raw", "Classified"]
THRESHOLDS = [1, 2, 3, 4, 5]


def run_ci(values, weeks, post_start):
    df = pd.DataFrame({"y": values}, index=weeks).astype(float)
    pre_end = post_start - pd.Timedelta(days=7)
    post_end = post_start + pd.Timedelta(days=7 * 7)
    ci = CausalImpact(df, [df.index.min(), pre_end], [post_start, post_end], alpha=0.05)
    s = ci.summary()
    rel = p = None
    for line in s.split("\n"):
        if "Relative effect" in line and rel is None:
            for tok in line.split():
                try: rel = float(tok.strip("%()").replace(",", "")); break
                except: pass
        if "tail-area" in line:
            for tok in line.split():
                try:
                    v = float(tok)
                    if 0 <= v <= 1: p = v
                except: pass
    return rel, p


if __name__ == "__main__":
    path = os.path.join(DATA_DIR, "table17_weekly_series.csv")
    if not os.path.exists(path):
        print(f"ERROR: {path} not found.")
        print("Copy table17_weekly_series.csv from your results/table17_independent/ directory.")
        exit(1)

    weekly = pd.read_csv(path)
    weekly["week"] = pd.to_datetime(weekly["week"])
    weekly["k"] = weekly["k"].astype(int)

    total = len(GROUPS) * len(DEFINITIONS) * len(THRESHOLDS) * len(MILESTONES)
    print(f"Running {total} Table 17 models...\n")

    rows = []
    count = 0
    for group in GROUPS:
        for defn in DEFINITIONS:
            for k in THRESHOLDS:
                sub = weekly[(weekly["group"] == group) & (weekly["definition"] == defn) &
                             (weekly["k"] == k)].sort_values("week")
                if len(sub) == 0: continue
                for ms_name, ms_date in MILESTONES.items():
                    count += 1
                    print(f"[{count:2d}/{total}] {group:8s} {defn:10s} k>={k} {ms_name}", end=" ", flush=True)
                    try:
                        rel, p = run_ci(sub["total"], sub["week"], ms_date)
                        sig = "**" if p and p < 0.05 else ("*" if p and p < 0.10 else "")
                        print(f"{rel:+.0f}%{sig}" if rel else "ERROR")
                    except:
                        rel, p = None, None
                        print("ERROR")
                    rows.append({"group": group, "definition": defn, "k": k,
                                 "milestone": ms_name,
                                 "relative_effect_pct": round(rel, 1) if rel else None,
                                 "p_value": round(p, 4) if p else None})

    out = pd.DataFrame(rows)
    out.to_csv(os.path.join(DATA_DIR, "table17_reproduced.csv"), index=False)
    print(f"\nSaved: {DATA_DIR}/table17_reproduced.csv ({len(out)} rows)")

File: rq1/scripts/table_17.R
library(CausalImpact)
library(zoo)

DATA_DIR <- "rq1/data"
MILESTONES <- list(Oct2023 = as.Date("2023-10-23"), Mar2025 = as.Date("2025-03-17"), Jul2025 = as.Date("2025-07-21"))

snap_to_week <- function(target, weeks) {
  candidates <- weeks[weeks >= target]
  if (length(candidates) > 0) return(min(candidates))
  return(max(weeks[weeks < target]))
}

run_ci <- function(y, weeks, treatment_target) {
  td <- snap_to_week(treatment_target, weeks)
  pre_end <- td - 7; post_end <- td + 7 * 7
  idx <- weeks >= min(weeks) & weeks <= post_end
  s <- zoo(y[idx], order.by = weeks[idx])
  if (length(unique(y[idx])) <= 1) return(c(NA, NA))
  set.seed(20260827)
  tryCatch({
    ci <- CausalImpact(s, c(min(weeks), pre_end), c(td, post_end))
    avg <- ci$summary["Average", ]
    c(round(avg[["RelEffect"]] * 100, 1), round(avg[["p"]], 4))
  }, error = function(e) c(NA, NA))
}

weekly <- read.csv(file.path(DATA_DIR, "table17_weekly_series.csv"), stringsAsFactors = FALSE, check.names = FALSE)
weekly$week <- as.Date(weekly$week)
weekly$k <- as.integer(weekly$k)

total <- 60
cat(sprintf("Running %d Table 17 models...\n\n", total))

results <- data.frame(); count <- 0
for (grp in c("VPN", "Politics")) {
  for (defn in c("Raw", "Classified")) {
    for (k in 1:5) {
      sub <- weekly[weekly$group == grp & weekly$definition == defn & weekly$k == k, ]
      sub <- sub[order(sub$week), ]
      if (nrow(sub) == 0) next
      for (ms in names(MILESTONES)) {
        count <- count + 1
        cat(sprintf("[%2d/%d] %s | %s | k>=%d | %s ... ", count, total, grp, defn, k, ms))
        r <- run_ci(sub$total, sub$week, MILESTONES[[ms]])
        if (!is.na(r[1])) {
          sig <- ifelse(r[2] < 0.05, "**", ifelse(r[2] < 0.10, "*", ""))
          cat(sprintf("%+.1f%%%s\n", r[1], sig))
        } else cat("NA\n")
        results <- rbind(results, data.frame(group = grp, definition = defn, k = k,
          milestone = ms, relative_effect_pct = r[1], p_value = r[2]))
      }
    }
  }
}
write.csv(results, file.path(DATA_DIR, "table17_reproduced.csv"), row.names = FALSE)
cat(sprintf("\nSaved: table17_reproduced.csv (%d rows)\n", nrow(results)))
